File: polymorphism_and_monomorphism/mono.cc
#include <stdio.h>

struct A
{
};
struct B
{
};

template <typename T>
struct Generic
{
    static void fn();
};

template <>
struct Generic<A>
{
    static void fn()
    {
        printf("foo\n");
    }
};

template <>
struct Generic<B>
{
    static void fn()
    {
        printf("bar\n");
    }
};

template <typename T>
static void call(T obj)
{
    Generic<T>::fn();
}

int main()
{
    struct A a;
    struct B b;
    call(a);
    call(b);
}

File: polymorphism_and_monomorphism/mono.rs
trait Generic {
    fn foo(&self);
}

struct A;
struct B;

impl Generic for A {
    fn foo(&self) {
        println!("A")
    }
}

impl Generic for B {
    fn foo(&self) {
        println!("B")
    }
}

fn call(bar: &impl Generic) {
    bar.foo()
}

fn main() {
    call(&A {});
    call(&B {});
}
